pkg/log/zap/flags.go

```diff
@@ -78,11 +80,17 @@ func (v encoderValue) Type() string {
 
 func jsonEncoder() zapcore.Encoder {
 	encoderConfig := zap.NewProductionEncoderConfig()
+	if timeformatVal.set {
+		encoderConfig.EncodeTime = timeformatVal.timeEncoder
+	}
```

Comment:
As I mentioned in my [earlier review comment](https://github.com/operator-framework/operator-sdk/pull/1529#pullrequestreview-247092609), this will cause problems when `--zap-encoder` is defined before `--zap-timeformat`, because `jsonEncoder()` (or `consoleEncoder()`) will be called before `timeformatVal` is set.

```console
$ operator-sdk up local --operator-flags="--zap-encoder=json --zap-timeformat=nanos"
INFO[0000] Running the operator locally.                
INFO[0000] Using namespace default.                  
{"level":"info","ts":1560389746.122705,"logger":"cmd","msg":"Go Version: go1.12.4"}
{"level":"info","ts":1560389746.123007,"logger":"cmd","msg":"Go OS/Arch: darwin/amd64"}
{"level":"info","ts":1560389746.123085,"logger":"cmd","msg":"Version of operator-sdk: v0.8.0+git"}
{"level":"info","ts":1560389746.124311,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1560389746.124352,"logger":"leader","msg":"Skipping leader election; not running in a cluster."}
{"level":"info","ts":1560389746.2729468,"logger":"cmd","msg":"Registering Components."}
{"level":"info","ts":1560389746.357215,"logger":"cmd","msg":"failed to initialize service object for metrics: OPERATOR_NAME must be set"}
{"level":"info","ts":1560389746.357249,"logger":"cmd","msg":"Starting the Cmd."}
```

As you can see, the `nanos` time format is not honored.

To fix this, I think we need to do some refactoring such that the encoder is created only after all of the flags have been parsed. To make this extensible for other future encoder options, I think a good option would be to refactor the encoder to be created with the [functional options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

We could refactor `encoderVal` and `jsonEncoder` (and similarly `consoleEncoder`), and add a new `encoderConfigFunc` called `withTimeFormat` in `flags.go`:

```go
type encoderConfigFunc func(*zapcore.EncoderConfig)

type encoderVal struct {
	set        bool
	newEncoder func(...encoderConfigFunc) zapcore.Encoder
	str        string
}

func (v *encoderValue) Set(e string) error {
	...
	case "json":
		v.newEncoder = newJSONEncoder
	...
	return nil
}

...

func newJSONEncoder(ecfs ...encoderConfigFunc) zapcore.Encoder {
	encoderConfig := zap.NewProductionEncoderConfig()
	for _, f := range ecfs {
		f(&encoderConfig)
	}
	return zapcore.NewJSONEncoder(encoderConfig)
}

...

func withTimeFormat(te zapcore.TimeEncoder) encoderConfigFunc {
	return func(ec *zapcore.EncoderConfig) {
		ec.EncodeTime = te
	}
}
```

And then in `logger.go` we could update `getConfig()` to look like:
```go
func getConfig() config {
	...

	var newEncoder func(...encoderConfigFunc) zapcore.Encoder
	if development {
		newEncoder = newConsoleEncoder
		...
	} else {
		newEncoder = newJSONEncoder
		...
	}

	var ecfs []encoderConfigFunc
	if encoderVal.set {
		newEncoder = encoderVal.newEncoder
	}

	...

	if timeformatVal.set {
		ecfs = append(ecfs, withTimeFormat(timeformatVal.timeEncoder))
	}
	c.encoder = newEncoder(ecfs...)

	...

	return c
}
```

Does this make sense? I know this suggestion is a bigger refactor than you were probably expecting, so feel free to ask me to explain things more.

Reply:
I think I see what you're seeing (and sorry for missing it the first time). 

To be honest this approach is a bit beyond my current skill level with Go but I'll try and see if I can implement this as you've proposed and come back with something soon (hopefully!)

Reply:
Alright I think I've incorporated all of that. Testing on my side looks OK, but again I'm a bit out of my depth here so I appreciate any other feedback you can offer?

Also, you basically wrote everything and I copy-and-pasted most. Thanks for that, too :D